ci_framework/roles/override_op_image/tasks/main.yml

```diff
@@ -0,0 +1,52 @@
+---
+# Copyright Red Hat, Inc.
+# All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Ensure directory exists
+  ansible.builtin.file:
+    path: "{{ cifmw_override_op_image_dir }}/{{ item }}"
+    state: directory
+  loop:
+    - artifacts/manifests/openstack
+    - logs
+
+- name: Update CSVs with custom images
+  when:
+  - cifmw_openstack_svc_image is defined
+  - cifmw_openshift_kubeconfig is defined
+  environment:
+    KUBECONFIG: "{{ cifmw_openshift_kubeconfig }}"
+    PATH: "{{ cifmw_path }}"
+  block:
+    - name: Dump OpenStackControlPlane CR into a file
+      ci_script:
+        output_dir: "{{ cifmw_override_op_image_dir }}/artifacts/manifests/openstack"
+        script: |
+          cp_name=$(oc get OpenStackControlPlane -n {{ cifmw_install_yamls_defaults['NAMESPACE'] }} --no-headers=true | awk '{print $1}')
+          oc get OpenStackControlPlane { cp_name } -n {{ cifmw_install_yamls_defaults['NAMESPACE'] }} -o yaml > {{ cifmw_override_op_image_dir }}/artifacts/manifests/openstack/osp_cp_cr.yaml
+
+    - name: Replace openstack service images in the yaml file
+      with_dict: "{{cifmw_openstack_svc_image}}"
+      ansible.builtin.replace:
+        path: "{{ cifmw_override_op_image_dir }}/artifacts/manifests/openstack/osp_cp_cr.yaml"
+        regexp: "{{ item.key }}\\[[^\\]]+\\]"
```

Comment:
This is the regex I said I don't really understand. Anyway, can't this be done in Ansible instead of manipulating the manifest based on regexes?

Reply:
The file containing the images, looks something like this:
```
cifmw_openstack_svc_image:
  quay.io/podified-antelope-centos9/openstack-ceilometer-central: sha256:5904247ce6322e2fb56830572e868e6d2937490ab7167865be5fb01d3f1f780d
  quay.io/podified-antelope-centos9/openstack-ceilometer-notification: sha256:8e3b56862a5f047b799edaa30a8e0b342b402f5181d397120327d3d9c4272564
  quay.io/infrawatch/sg-core: sha256:4befc0268975239bf67f895d9a5b9304e4733182e61ece527aba53e55582883d
  quay.io/podified-antelope-centos9/openstack-cinder-api: sha256:c8cfce053db833ba4ff141598e281213d001f14b6f23340caa0c140e7f58db3c
```
The regex searches for any string in the CR that matches the key (which also happens to be the image name) and replaces that with image name (key) and image_tag (value)  from the file.